fabric-object-builder-api-v1/src/main/java/net/fabricmc/fabric/impl/object/builder/TradeOfferInternals.java

```diff
@@ -29,94 +27,29 @@
-		final List<TradeOffers.Factory> list = new ArrayList<>();
-		factory.accept(list);
-
-		final TradeOffers.Factory[] additionalEntries = list.toArray(new TradeOffers.Factory[0]);
-		final TradeOffers.Factory[] currentEntries = TradeOfferInternals.DEFAULT_WANDERING_TRADER_OFFERS.computeIfAbsent(level, key -> new TradeOffers.Factory[0]);
-
-		// Merge current and new entries
-		final TradeOffers.Factory[] newEntries = ArrayUtils.addAll(additionalEntries, currentEntries);
-		TradeOfferInternals.DEFAULT_WANDERING_TRADER_OFFERS.put(level, newEntries);
-
-		// Refresh the trades map
-		TradeOfferInternals.refreshOffers();
+		registerOffers(TradeOffers.WANDERING_TRADER_TRADES, level, factory);
 	}
 
-	public static void refreshOffers() {
-		TradeOfferInternals.refreshVillagerOffers();
-		TradeOfferInternals.refreshWanderingTraderOffers();
-	}
-
-	private static void refreshVillagerOffers() {
-		final HashMap<VillagerProfession, Int2ObjectMap<TradeOffers.Factory[]>> trades = new HashMap<>(TradeOfferInternals.DEFAULT_VILLAGER_OFFERS);
-
-		for (Map.Entry<VillagerProfession, Int2ObjectMap<TradeOffers.Factory[]>> tradeFactoryEntry : TradeOfferInternals.VILLAGER_TRADE_FACTORIES.entrySet()) {
-			// Create an empty map or get all existing profession entries.
-			final Int2ObjectMap<TradeOffers.Factory[]> leveledFactoryMap = trades.computeIfAbsent(tradeFactoryEntry.getKey(), k -> new Int2ObjectOpenHashMap<>());
-			// Get the existing entries
-			final Int2ObjectMap<TradeOffers.Factory[]> value = tradeFactoryEntry.getValue();
-
-			// Iterate through the existing level entries
-			for (int level : value.keySet()) {
-				final TradeOffers.Factory[] factories = value.get(level);
-
-				if (factories != null) {
-					final Int2ObjectMap<TradeOffers.Factory[]> resultMap = trades.computeIfAbsent(tradeFactoryEntry.getKey(), key -> new Int2ObjectOpenHashMap<>());
-					resultMap.put(level, ArrayUtils.addAll(leveledFactoryMap.computeIfAbsent(level, key -> new TradeOffers.Factory[0]), factories));
-				}
-			}
-		}
-
-		// Set the new villager trade map
-		TradeOffersAccessor.setVillagerTradeMap(trades);
-	}
-
-	private static void refreshWanderingTraderOffers() {
-		// Create an empty map that is a clone of the default offers
-		final Int2ObjectMap<TradeOffers.Factory[]> trades = new Int2ObjectOpenHashMap<>(TradeOfferInternals.DEFAULT_WANDERING_TRADER_OFFERS);
-
-		for (int level : TradeOfferInternals.WANDERING_TRADER_FACTORIES.keySet()) {
-			// Get all registered offers and add them to current entries
-			final TradeOffers.Factory[] factories = TradeOfferInternals.WANDERING_TRADER_FACTORIES.get(level);
-			trades.put(level, ArrayUtils.addAll(factories, trades.computeIfAbsent(level, key -> new TradeOffers.Factory[0])));
-		}
+	// Shared code to register offers for both villagers and wandering traders.
+	// synchronized guards against concurrent modifications - Vanilla does not mutate the underlying arrays (as of 1.16),
+	// so reads will be fine without locking.
```

Comment:
Not sure this is useful... In theory all registrations should happen during the init phase.

Reply:
Note to myself: the `computeIfAbsent` call on line 39 needs to be `synchronized` as well.